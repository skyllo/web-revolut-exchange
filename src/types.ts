export enum Pocket {
  EUR = 'EUR',
  GBP = 'GBP',
  USD = 'USD',
}

export interface PocketData {
  name: Pocket;
  balance: number;
  symbol: string;
  title: string;
  countryCode: string;
}

export type InputDirection = 'from' | 'to';

export interface Rate {
  rate: number;
  fromPocket: Pocket;
  toPocket: Pocket;
}
