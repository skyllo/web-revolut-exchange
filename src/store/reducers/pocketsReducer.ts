/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { POCKET_TO_SYMBOL } from '../../utils/constants';
import { Pocket, PocketData } from '../../types';

type PocketAction = PayloadAction<{
  pocket: Pocket;
  balance: number;
}>;

export const initialStatePockets: Record<Pocket, PocketData> = {
  EUR: {
    name: Pocket.EUR,
    balance: 100,
    symbol: POCKET_TO_SYMBOL[Pocket.EUR],
    title: 'Euro',
    countryCode: 'EU',
  },
  GBP: {
    name: Pocket.GBP,
    balance: 200,
    symbol: POCKET_TO_SYMBOL[Pocket.GBP],
    title: 'British Pound',
    countryCode: 'GB',
  },
  USD: {
    name: Pocket.USD,
    balance: 300,
    symbol: POCKET_TO_SYMBOL[Pocket.USD],
    title: 'US Dollar',
    countryCode: 'US',
  },
};

export const pocketsSlice = createSlice({
  name: 'pockets',
  initialState: initialStatePockets,
  reducers: {
    setBalance: (state, { payload: { balance, pocket } }: PocketAction) => {
      state[pocket].balance = balance;
    },
  },
});

export const { setBalance } = pocketsSlice.actions;

export default pocketsSlice.reducer;
