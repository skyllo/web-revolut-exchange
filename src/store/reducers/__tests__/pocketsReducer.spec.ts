import { Pocket } from '../../../types';
import pocketsReducer, { initialStatePockets, setBalance } from '../pocketsReducer';

describe('pocketsReducer', () => {
  it('should return the initial state', () => {
    expect(pocketsReducer(undefined, { type: '' })).toMatchSnapshot();
  });

  it('should set balance correctly', () => {
    expect(
      pocketsReducer(
        {
          EUR: {
            ...initialStatePockets.EUR,
            balance: 0,
          },
          GBP: {
            ...initialStatePockets.GBP,
            balance: 0,
          },
          USD: {
            ...initialStatePockets.USD,
            balance: 0,
          },
        },
        setBalance({ balance: 199, pocket: Pocket.EUR })
      )
    ).toMatchSnapshot();
  });
});
