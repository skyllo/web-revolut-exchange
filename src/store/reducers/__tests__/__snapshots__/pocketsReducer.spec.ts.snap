// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pocketsReducer should return the initial state 1`] = `
Object {
  "EUR": Object {
    "balance": 100,
    "countryCode": "EU",
    "name": "EUR",
    "symbol": "€",
    "title": "Euro",
  },
  "GBP": Object {
    "balance": 200,
    "countryCode": "GB",
    "name": "GBP",
    "symbol": "£",
    "title": "British Pound",
  },
  "USD": Object {
    "balance": 300,
    "countryCode": "US",
    "name": "USD",
    "symbol": "$",
    "title": "US Dollar",
  },
}
`;

exports[`pocketsReducer should set balance correctly 1`] = `
Object {
  "EUR": Object {
    "balance": 199,
    "countryCode": "EU",
    "name": "EUR",
    "symbol": "€",
    "title": "Euro",
  },
  "GBP": Object {
    "balance": 0,
    "countryCode": "GB",
    "name": "GBP",
    "symbol": "£",
    "title": "British Pound",
  },
  "USD": Object {
    "balance": 0,
    "countryCode": "US",
    "name": "USD",
    "symbol": "$",
    "title": "US Dollar",
  },
}
`;
